swap_meet/vendor.py

```diff
@@ -1,2 +1,160 @@
+
+    swap_items(other, my_item, their_item):
+        Swaps provided items between two instances' inventories and returns True.
+        - Returns False if either item not found.
+
+    swap_first_item(other):
+        Swaps first items of two instances' inventories and returns True.
+        - Returns false if either inventory is empty.
+
+    get_best_by_category(desired_category):
+        Returns item of provided category with best condition from an instance's inventory. 
+        - Returns None if no items of provided category in inventory.
+
+    swap_best_by_category(other, my_priority, their_priority):
+        Swaps best items in provided categories between two instances' inventories and returns True.
+        - Returns False if either item not found. 
+
+    get_newest_item():
+        Returns newest item from instance's inventory items with known ages.
+        - Returns None if all items ages are unknown.
+
+    swap_by_newest(other):
+        Swaps newest items of two instances' inventories and returns True.
+        - Returns False if either or both inventory items' ages are all unknown.
+    
+    """
+
+    def __init__(self, inventory = None):
+        if inventory is None:
+            inventory = []
+        self.inventory = inventory
+    
+    def add(self, new_item):
+        """Add provided item to inventory"""
+
+        self.inventory.append(new_item)
+        return new_item
+    
+    def remove(self, discarded_item):
+        """Remove an item from inventory"""
+
+        if discarded_item not in self.inventory:
+            return False
+        else: 
+            self.inventory.remove(discarded_item)
+            return discarded_item
+    
+    def get_by_category(self, category_string):
+        """Return list of items from inventory of a certain category"""
+
+        item_list = []
+        for item in self.inventory:
+            if item.category == category_string:
+                item_list.append(item)
+        return item_list
+    
+    def swap_items(self, other, my_item, their_item):
+        """Swaps items betwen self and another vendor's inventory"""
+
+        # check valid input
+        if my_item not in self.inventory or their_item not in other.inventory:
+            return False
+        else:
+            # remove items from original inventory
+            self.remove(my_item)
+            other.remove(their_item)
+            # add items to new inventory 
+            self.add(their_item)
+            other.add(my_item)
+            return True
+    
+    def swap_first_item(self, other):
+        """Swaps first item in self and another vendor's inventory"""
+
+        # check valid input
+        if not self.inventory or not other.inventory:
+            return False
+        else:
+            my_first_item = self.inventory[0]
+            vendor_first_item = other.inventory[0]
+            self.swap_items(other, my_first_item, vendor_first_item)
```

Comment:
This whole tuple swap thing that came up here in class as well is something I'd like to talk about at our next meeting! I'm just so perplexed as to what is happening under the hood that python knows what to do with just this expression and it's still confusing to me. I'd like to understand it better!

Reply:
Definitely, let's talk it through! High-level, what's happening is python is assigning multiple variables at the same time (there are many languages that allow this). We could use the same syntax if we need to assign multiple strings or any other type at the same time:
```python
# After this line, var_a will point to "value 1" and var_b will point to "value 2"
var_a, var_b = "value 1", "value 2" 

# After this line, int_a will point to 5, int_b will point to 12, and int_c will point to 20
int_a, int_b, int_c = 5, 12, 20
```
As you can see in the int example above, we're also not limited to only assigning 2 variables at a given time. We often choose not to do this multiple assignment because it can make code less readable by making it harder to see where an individual variable is declared, but for a direct swap operation like we're solving in Swap-Meet, it becomes really handy!

If we look back at the Swap-Meet code:
- On the left side we declare the variables we are assigning, `self.inventory[0]` and  `other.inventory[0]`. In this case instead of named variables they are locations in our vendor inventory lists, which like named variables are pointers to data. 

- On the right side we say what values the variables are being assigned. In this case, instead of a raw piece of data or a named variable, those values are the data pointed to by the first indices in our vendor inventories.

As I understand it, python will look up the values to be assigned on the right of the `=` operator, then perform the assignment of those values to the variables on the left of the `=`.

Reply:
> As I understand it, python will look up the values to be assigned on the right of the = operator, then perform the assignment of those values to the variables on the left of the =.
This makes sense thank you!!!! That was, I think, the phrasing I needed.
And this is what happens instantaneously, I'm guessing, so we don't get a similar issue as if we were changing a list we were iterating on? It's not like we change self.inventory[0] and then when we look there to change other.inventory[0] it's the new value - It checks the original values of both locations, and then assigns the new locations to those values?

Reply:
Exactly! If python instead tried to resolve the multiple assignments one after the other, in Swap-Meet we'd see a bug where the first list would get changed as we expect, but the second list would appear to not have been updated.

Reply:
If you're interested, this blog has some really nice examples of how wild multiple assignment can get and different ways it can be used: https://note.nkmk.me/en/python-multi-variables-values/

Reply:
Thank you!!! This was super helpful. Sometimes you just need something phrased in a certain way for it to click!!